### PR TITLE
fix(build): make Node.js scripts work with Node.js 20

### PR DESCRIPTION
**Pull request by maxpatiiuk**
- [x] Don't hardcode Node.js version number in the GitHub Action Workflows, but get it dynamically from the engines field in the package.json
- [x] Update GitHub Action to newest versions
- [x] Run prettier on the .yml files

### Checklist

- [x] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [x] Add automated tests

### Testing instructions

Most changes in this PR are build-related, so don't affect the user-facing features, as long as everything builds correctly. Still, to be safe, please verify that map in query builder works correctly:

- [ ] Leaflet map shows up correctly in the following places:
  - [ ] When entering Lat/long in query builder, can open the map and specify the coordinates using the ap
  - [ ] If query includes locality table, can plot results on the map
- [x] All GitHub Actions should pass, and use the correct version:
 Screenshot 2024-02-19 at 10 29 24

### CODE REVIEW

File: .github/workflows/codeql.yml
Comment (maxpatiiuk) [inline .github/workflows/codeql.yml:1]: formatting changes only

**Comment by maxpatiiuk** [comment]
I also made an analogous change on the `weblate-localization` branch in 9b2580111 to fix the GitHub Action failing https://github.com/specify/specify7/actions/runs/7962942939/job/21737510948
Screenshot 2024-02-19 at 10 31 30

cc @acwhite211

**Review by CarolineDenis — APPROVED** [review]

**Comment by maxpatiiuk** [comment]
After spending several hours trying to resolve the build issues from https://github.com/specify/specify7/actions/runs/7963889013/job/21740345853, I got these results:

- As a workaround for https://github.com/privatenumber/tsx/issues/38, converted .js files to .ts
   But also, it's time for those to be TypeScript
   Note, I added `// @ts-nocheck` to the top of the files to suppress
   TypeScript errors for now, but this should be removed once we move way from Backbone
- Refactored our configs to use ESM over Common JS because I wasn't able to get swc-node to work.
   tsx however works fine when using "type": "module" in package.json.
   And it's about time to convert to ESM anyway - while there would still be some friction expected, going forward, there would be more friction from not using ESM than using ESM
- Fix issues with Leaflet in node.js. See commit comment in https://github.com/specify/specify7/pull/4558/commits/342a727f5b1cc63b037e64d0ace54b2e37b63d17

Outcome:
- We are now using ESM, with some dependencies updated to newest version, and no more .js files
- We are on Node.js 20.11, and everything works

File: specifyweb/frontend/js_src/lib/components/DataModel/backbone.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+// @ts-nocheck
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/DataModel/backbone.ts:1]: no code changes - just ran prettier on this file

File: specifyweb/frontend/js_src/package.json [hunk 2]
```diff
@@ -111,12 +110,13 @@
...
     "postcss": "^8.4.14",
-    "postcss-loader": "^7.0.1",
+    "postcss-loader": "^8.1.0",
     "postcss-preset-env": "^7.7.2",
     "prettier": "^2.7.1",
     "regenerator-runtime": "^0.13.9",
-    "style-loader": "^3.3.1",
-    "tailwindcss": "^3.2.4",
+    "style-loader": "^3.3.4",
+    "tailwindcss": "^3.4.1",
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/package.json:118]: tailwind updated to get ESM+TS support - no breaking changes

File: specifyweb/frontend/js_src/tailwind.config.ts [hunk 3]
```diff
@@ -3,15 +3,21 @@
...
  */
 
+import type {Config} from 'tailwindcss';
+import forms from '@tailwindcss/forms';
+
 /*
+ * REFACTOR: https://tailwindcss.com/blog/tailwindcss-v3-3#css-variables-without-the-var
+ * REFACTOR: https://tailwindcss.com/blog/tailwindcss-v3-4#style-children-with-the-variant
+ * REFACTOR: https://tailwindcss.com/blog/tailwindcss-v3-4#subgrid-support
+ * REFACTOR: https://tailwindcss.com/blog/tailwindcss-v3-4#extended-min-width-max-width-and-min-height-scales
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/tailwind.config.ts:13]: nice improvements, but didn't want to introduce any bugs in this PR, so will delay refactor to some future point

File: specifyweb/frontend/js_src/tsconfig.json [hunk 4]
```diff
@@ -44,8 +44,7 @@
     "noEmit": true,
     "sourceMap": true,
     "target": "esnext",
-    "allowJs": true,
-    "checkJs": false,
+    "allowJs": false,
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/tsconfig.json:47]: 🎉

**Review by CarolineDenis — APPROVED** [review]

**Review by pashiav — APPROVED** [review]
Works as expected!

**Review by melton-jason — APPROVED** [review]
Glad to see this issue finally resolved! 

To better ensure that the suggested changes (properly convert the old js files to ts and the REFACTOR suggestions with tailwind) are not lost and forever stuck as comments, the context can be put into an Issue (or two), if there is not another Issue which can capture the required context.

The Javascript files here are near their end! They've definitely overstayed their welcome 😄

**Pull request merged**